File: utils/combat/resolver.js
const types = require('../data/types');

/**
 * Combat Resolver: Handles refined GDD damage calculation.
 * Damage = (BaseMoveDamage + (AttackPower * 0.2)) * TypeMultiplier * CriticalMultiplier * ResilienceMultiplier
 */
module.exports = {
    resolveAttack(attacker, defender, move, isMeterFull = false) {
        // CE Check
        if (attacker.ce < move.ce) {
            return {
                damage: 10, // Default "Basic Punch"
                isBlackFlash: false,
                msg: `${attacker.name} tried to use ${move.name} but lacked Cursed Energy!`
            };
        }

        // 1. Base Damage Calculation
        const [min, max] = move.dmg || [10, 20];
        const baseMoveDmg = Math.floor(Math.random() * (max - min + 1)) + min;
        const attackPower = Number(attacker.atk || attacker.attack || 50);

        // 2. Multipliers
        const typeMult = Number(types.getMultiplier(move.type, defender.type) || 1.0);
        
        // --- Technical Logic Expansion ---
        let resilience = Number(defender.resilience || 0);
        if (move.ignoreBarrier || (move.effect && move.effect.ignoreBarrier)) {
            resilience = 0; // Completely bypass resilience
        }
        
        let resilienceMult = 1 - (resilience / 100);
        
        // Defense Piercing logic
        let piercer = 1.0;
        if (move.ignoreDef || (move.effect && move.effect.ignoreDef)) {
            piercer = 1.5; // Acts as a damage boost if defense is ignored
        }

        // 3. Black Flash (Crit)
        let bfChance = Number(move.crit || 10) / 100;
        
        if (attacker.hasBFBuff) {
            bfChance += 0.05;
        }

        let isBlackFlash = Math.random() < bfChance || isMeterFull;
        let critMult = 1.0;
        
        if (isBlackFlash) {
            critMult = isMeterFull ? 3.0 : 2.0;
        }

        // --- Accuracy & Evasion Checks ---
        const moveAcc = move.accuracy || 100;
        const defEvasion = Math.min(25, (defender.speed || 100) / 40); // Max 25% evasion
        
        let isMiss = Math.random() * 100 > moveAcc;
        let isDodge = !isMiss && Math.random() * 100 < defEvasion;
        
        if (isMiss || isDodge) {
            return {
                damage: 0,
                isBlackFlash: false,
                ceCost: move.ce,
                isMiss,
                isDodge,
                typeMult: 1.0,
                isAOE: false,
                meterGain: 0
            };
        }

        // 4. Final Calculation
        // Increased attack multiplier from 0.2 to 1.5 for faster, exciting combat
        let damage = Math.floor((baseMoveDmg + (attackPower * 1.5)) * typeMult * critMult * resilienceMult * piercer);
        if (isNaN(damage)) damage = 10;

        // 5. Positioning Modifier
        const attackerPos = attacker.position || 'middle';
        const defenderPos = defender.position || 'middle';
        const atkPosMult = attackerPos === 'front' ? 1.1 : (attackerPos === 'back' ? 0.95 : 1.0);
        const defPosMult = defenderPos === 'front' ? 1.2 : (defenderPos === 'back' ? 0.9 : 1.0);
        
        damage = Math.floor(damage * atkPosMult * (1 / defPosMult));

        // 6. Random Variance (85% to 100% - Pokemon style)
        const variance = (Math.floor(Math.random() * 16) + 85) / 100;
        damage = Math.floor(damage * variance);

        let typeMsg = "";
        if (typeMult > 1.2) typeMsg = "💥 <b>It's Super Effective!</b>";
        else if (typeMult < 0.8) typeMsg = "🛡️ <b>It's not very effective...</b>";

        return {
            damage,
            isBlackFlash,
            ceCost: move.ce,
            typeMult,
            typeMsg,
            isAOE: move.aoe || false,
            meterGain: isBlackFlash ? 20 : 0
        };
    }
};

File: utils/combat/resolver.py
import random
from utils.data.types import types_system

class CombatResolver:
    def resolve_attack(self, attacker, defender, move, is_meter_full=False):
        # CE Check
        if attacker.get('ce', 0) < move.get('ce', 0):
            return {
                "damage": 10,
                "isBlackFlash": False,
                "msg": f"{attacker['name']} tried to use {move['name']} but lacked Cursed Energy!"
            }

        # 1. Base Damage Calculation
        min_dmg, max_dmg = move.get('dmg', [10, 20])
        base_move_dmg = random.randint(min_dmg, max_dmg)
        
        power = float(attacker.get('power', 50))
        stamina = float(attacker.get('stamina', 50))
        ce_stat = float(attacker.get('ce_stat', 50))
        technique = float(attacker.get('technique', 50))
        level = float(attacker.get('level', 1))
        
        # Power Metrics Influence
        a_tp = float(attacker.get('tp', 100))
        a_ts = float(attacker.get('ts', 100))
        d_tp = float(defender.get('tp', 100))
        d_ts = float(defender.get('ts', 100))

        # TS adds a flat multiplier to the character's potency
        ts_mult = 1.0 + (a_ts / 2500.0)
        # TP adds to the character's overall stability and secondary scaling
        tp_mult = 1.0 + (a_tp / 5000.0)

        # Tank Build Scaling: 0.2 to 6.0 based on Level
        scale = 0.2 + (level / 20.0)
        if scale > 6.0: scale = 6.0

        is_cursed = move.get('type') in ['Long-range', 'Barrier', 'Cursed']
        
        if is_cursed:
            # Cursed Damage Scaling
            attack_power = ((ce_stat * 1.5) + (technique * 0.8)) * ts_mult
        else:
            # Physical Damage Scaling
            attack_power = ((power * 0.8) + (stamina * 0.1)) * ts_mult

        # 2. Multipliers
        type_mult = types_system.get_multiplier(move.get('type'), defender.get('type'))
        
        # Resilience (Defense scaling) - Boosted by Defender's TP
        resilience_base = float(defender.get('stamina', 50)) / 10.0
        resilience_tp_bonus = d_tp / 100.0
        resilience = resilience_base + resilience_tp_bonus
        
        if move.get('ignoreBarrier') or move.get('effect', {}).get('ignoreBarrier'):
            resilience = 0
            
        resilience_mult = 100.0 / (100.0 + resilience)
        
        # Piercing
        piercer = 1.0
        if move.get('ignoreDef') or move.get('effect', {}).get('ignoreDef'):
            piercer = 1.5

        # 3. Black Flash (Crit) - Precision influenced by Technique and TS
        bf_chance = (float(move.get('crit', 10)) + (technique / 15.0) + (a_ts / 50.0)) / 100
        if attacker.get('hasBFBuff'):
            bf_chance += 0.05

        is_black_flash = random.random() < bf_chance or is_meter_full
        crit_mult = 1.0
        if is_black_flash:
            crit_mult = 3.0 if is_meter_full else 2.5 # Buffed BF mult

        # Accuracy & Evasion - Speed influenced by TS
        move_acc = move.get('accuracy', 100) + (a_ts / 20.0)
        def_evasion = min(45, (float(defender.get('speed', 100)) / 20.0) + (d_ts / 40.0))
        
        is_miss = random.random() * 100 > move_acc
        is_dodge = not is_miss and random.random() * 100 < def_evasion
        
        if is_miss or is_dodge:
            return {
                "damage": 0,
                "isBlackFlash": False,
                "ceCost": move.get('ce', 0),
                "isMiss": is_miss,
                "isDodge": is_dodge,
                "typeMult": 1.0,
                "isAOE": False,
                "meterGain": 0
            }

        # 4. Final Calculation
        damage = int((base_move_dmg + attack_power) * type_mult * crit_mult * resilience_mult * piercer * tp_mult)
        if damage < 1: damage = max(1, int(power * 0.05)) # Minimum damage based on Power

        # 5. Positioning
        attacker_pos = attacker.get('position', 'middle')
        defender_pos = defender.get('position', 'middle')
        atk_pos_mult = 1.1 if attacker_pos == 'front' else (0.95 if attacker_pos == 'back' else 1.0)
        def_pos_mult = 1.2 if defender_pos == 'front' else (0.9 if defender_pos == 'back' else 1.0)
        
        damage = int(damage * atk_pos_mult * (1.0 / def_pos_mult))

        # 6. Variance
        variance = random.randint(85, 100) / 100.0
        damage = int(damage * variance)

        type_msg = ""
        if type_mult > 1.2: type_msg = "💥 <b>It's Super Effective!</b>"
        elif type_mult < 0.8: type_msg = "🛡️ <b>It's not very effective...</b>"

        return {
            "damage": damage,
            "isBlackFlash": is_black_flash,
            "ceCost": move.get('ce', 0),
            "typeMult": type_mult,
            "typeMsg": type_msg,
            "isAOE": move.get('aoe', False),
            "meterGain": 20 if is_black_flash else 0
        }

combat_resolver = CombatResolver()
